airbyte-integrations/connectors/source-linnworks/source_linnworks/source.py

```diff
@@ -0,0 +1,324 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
+
+
+from abc import ABC, abstractmethod
+from datetime import date
+from typing import (Any, Iterable, List, Mapping, MutableMapping, Optional,
+                    Tuple, Union)
+from urllib.parse import parse_qsl, urlparse
+
+import pendulum
+import requests
+from airbyte_cdk.models.airbyte_protocol import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth.core import HttpAuthenticator
+from airbyte_cdk.sources.streams.http.requests_native_auth import \
+    Oauth2Authenticator
+from requests.auth import AuthBase
+
+from .util import normalize
+import json
+
+
+class LinnworksStream(HttpStream, ABC):
+    http_method = "POST"
+
+    @property
+    def url_base(self) -> str:
+        return self.authenticator.get_server()
+
+    def __init__(self, authenticator: Union[AuthBase, HttpAuthenticator] = None, start_date: str = None):
+        super().__init__(authenticator=authenticator)
+
+        self._authenticator = authenticator
+        self.start_date = start_date
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        if response.status_code != requests.codes.ok:
+            return None
+
+        json = response.json()
+        if not isinstance(json, list):
+            json = [json]
+        for record in json:
+            yield normalize(record)
+
+
+class LinnworksGenericPagedResult(ABC):
+    @abstractmethod
+    def paged_result(self, response: requests.Response) -> Mapping[str, Any]:
+        pass
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        result = self.paged_result(response)
+        del result["Data"]
+        result = normalize(result)
+
+        if result["page_number"] < result["total_pages"]:
+            result["page_number"] += 1
+            return result
+
+
+class Location(LinnworksStream):
+    # https://apps.linnworks.net/Api/Method/Locations-GetLocation
+    # Response: StockLocation https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Locations-ClassBase-StockLocation
+    # Allows 150 calls per minute
```

Comment:
It's great that you mention the rate limits but do you think you could incorporate these limits into backoffs / retries strategies? 🙏